fix(ops-platform): the ORM models and the Alembic migration build different schemas, so login and every config route fail

## Problem

The ops platform holds two schema owners that disagree.

The Alembic migration `0001_initial.py` builds one schema. The SQLAlchemy ORM
models in `src/shared/models/` declare a different schema. The API startup
path runs `Base.metadata.create_all`, so both owners run in the same
deployment.

The two schemas disagree on four points.

1. Table names. The migration creates `msps` and `organizations`. The models
   declare `msps` and `orgs`. The model foreign key targets `orgs.org_id`.
2. Primary keys. The migration gives `config_revisions` the key
   `(id, org_id)` with `id UUID`. The model gives it `revision_id BIGINT`
   with autoincrement plus `org_id`.
3. Partition strategy. The migration creates `config_revisions`,
   `device_status_snapshots`, and `audit_records` with `PARTITION BY HASH
   (org_id)` and 16 fixed hash partitions. The models declare
   `postgresql_partition_by: "LIST (org_id)"`. The model docstring at
   `config.py:30` still says "hash-partitioned by org_id", so the declaration
   and the docstring disagree too.
4. Partition creation. `ensure_org_partitions` runs `CREATE TABLE ...
   PARTITION OF ... FOR VALUES IN ('<uuid>')`. That statement is a LIST bound.
   PostgreSQL rejects a LIST bound on a table that is partitioned by hash.

`ensure_tables_exist` calls `create_all`, which uses `checkfirst`. `checkfirst`
skips a table that already exists. `create_all` never alters an existing
table, so it cannot repair the drift and it never reports the drift.

## Impact

The documented deployment runs `alembic upgrade head`. See
`mist-ops-platform/docs/operations.md:20`. After that command the database
holds the migration schema.

The API then starts and runs `create_all`. `create_all` skips every table that
the migration already created and creates the missing ORM-only tables, such as
`orgs`. The result is a mixed database that matches neither owner.

Two failures follow.

- The login route calls `ensure_org_partitions` for every organization of the
  caller. See `health.py:197`. The statement uses a LIST bound against a
  hash-partitioned parent. PostgreSQL raises an error. The call has no guard,
  so login returns a 500 response. Nobody can sign in.
- Every ORM query against `config_revisions` selects `revision_id`. The
  migration created no such column. The query fails with `UndefinedColumn`.
  The configuration routes, the audit routes, and the drift routes all fail.

## Risk

Critical. The platform cannot serve a request on a database that was built the
documented way. The defect blocks login, which blocks every other route. The
trigger needs no attacker and no unusual input. A normal deployment triggers
it.

## Evidence

`mist-ops-platform/migrations/versions/0001_initial.py:66`
```python
    # -- E-01 Organizations -----------------------------------------------
    op.create_table(
        "organizations",
        sa.Column("id", sa.Uuid(), primary_key=True),
```

`mist-ops-platform/src/shared/models/inventory.py:63`
```python
class Organization(Base, TimestampMixin):
    """Mist organization — primary tenant boundary."""

    __tablename__ = "orgs"

    org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
    )
```

`mist-ops-platform/migrations/versions/0001_initial.py:151`
```python
    op.execute(sa.text("""
            CREATE TABLE config_revisions (
                id          UUID NOT NULL,
                org_id      UUID NOT NULL REFERENCES organizations(id) ON DELETE CASCADE,
```
and `0001_initial.py:165`
```python
            ) PARTITION BY HASH (org_id)
            """))
    _create_hash_partitions("config_revisions")
```

`mist-ops-platform/src/shared/models/config.py:32`
```python
class ConfigRevision(Base):
    """Immutable configuration snapshot captured from Mist API."""

    __tablename__ = "config_revisions"
    __table_args__ = (
        UniqueConstraint("entity_id", "content_hash", "org_id", name="uq_revision_dedup"),
        {"postgresql_partition_by": "LIST (org_id)"},
    )

    revision_id: Mapped[int] = mapped_column(
        BigInteger,
        primary_key=True,
        autoincrement=True,
    )
    org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("orgs.org_id"),
        primary_key=True,
        nullable=False,
```

`mist-ops-platform/src/shared/db.py:82`
```python
async def ensure_org_partitions(engine, org_id: str) -> None:  # noqa: ANN001
    """Create LIST partitions for one org across all partitioned tables."""
    validated = uuid_mod.UUID(str(org_id))
    safe_suffix = validated.hex

    async with engine.begin() as conn:
        for table_name in PARTITIONED_TABLES:
            partition_name = f"{table_name}_org_{safe_suffix}"
            await conn.execute(
                text(
                    f"CREATE TABLE IF NOT EXISTS {partition_name} "
                    f"PARTITION OF {table_name} "
                    f"FOR VALUES IN ('{validated!s}')"
                )
            )
```

`mist-ops-platform/src/api/main.py:21`
```python
    from src.shared.db import ensure_tables_exist

    await ensure_tables_exist(engine)
```

## Reproduction

1. Start a PostgreSQL container for the platform.
2. Run `alembic upgrade head` in `mist-ops-platform/`. The command follows
   `docs/operations.md:20`.
3. Confirm the partition strategy. Run
   `psql -c "SELECT partstrat FROM pg_partitioned_table p JOIN pg_class c ON c.oid = p.partrelid WHERE c.relname = 'config_revisions'"`.
   The value is `h`, which means hash.
4. Start the API. `ensure_tables_exist` runs and reports success, because
   `checkfirst` skips every existing table.
5. Send `POST /api/v1/login` with a valid Mist API token.
6. Read the response. The route returns a 500 response. The server log holds
   the PostgreSQL error for the LIST bound on a hash-partitioned table.
7. Send `GET /api/v1/config/revisions?org_id=<uuid>`. The route fails, because
   the ORM selects `revision_id` and the table has no such column.

## Suggested solution

1. Pick one schema owner. Alembic is the correct owner, because the deployment
   documentation already runs `alembic upgrade head` and because Alembic can
   alter an existing table.
2. Remove the `ensure_tables_exist` call from the lifespan in
   `src/api/main.py`. Keep the function only for the test fixtures, or delete
   it.
3. Rewrite the ORM models so that every table name, column name, primary key,
   and foreign key matches the migration. Fix `orgs` against `organizations`,
   `revision_id` against `id`, and `orgs.org_id` against `organizations.id`.
4. Choose one partition strategy. If the platform keeps hash partitioning,
   delete `ensure_org_partitions` and the `ensure_org_partitions` call in
   `health.py`, because the 16 hash partitions already cover every
   organization. If the platform moves to LIST partitioning, write a new
   migration that rebuilds the three tables and update the model docstring at
   `config.py:30`.
5. Add a continuous integration check that starts a PostgreSQL service, runs
   `alembic upgrade head`, and then runs `alembic check`. The command reports
   any remaining drift between the models and the migration head.

## CI follow-up on #2031: the gate was red, and it is fixed

The first push turned the gate `pytest (ops platform)` red. Two items follow.
The failure was mine, not a break on `main`.

### Item 1: the failure and its cause

```
ERROR collecting tests/unit/shared/test_schema_agreement.py
tests/unit/shared/test_schema_agreement.py:23: in <module>
    from alembic.migration import MigrationContext
E   ModuleNotFoundError: No module named 'alembic'
=== 2 deselected, 1 error in 1.73s ===
```

The gate installs `mist-ops-platform/requirements-dev.txt` and nothing else.
`alembic>=1.14` sat in `mist-ops-platform/pyproject.toml` only. My new test
imports `alembic.migration` at module scope, so collection stopped and the whole
job exited with code 2.

### The fix for item 1

I added `alembic>=1.14` to `requirements-dev.txt`, with the reason in a comment.
That matches the style of the file, which documents every pin.

I did not use `pytest.importorskip`. A skip would turn the schema agreement test
into a no-op on the runner, and that would hide the exact defect this issue
reports.

### Item 2: the two deselected tests now run

`.github/workflows/ci.yml` deselected two tests and named this issue as the
repair. The repair landed, so the pull request removes the deselect list.

- Deleted both `--deselect` lines. No `--deselect` flag remains in the file.
- `TestOrganization::test_tablename` and `TestSyncLedgerEntry::test_tablename`
  both pass against the unified schema.
- Raised `OPS_PLATFORM_MIN_TESTS` from 330 to 390. The gate collected **398**
  tests on the merge result.
- Kept `OPS_PLATFORM_COVERAGE_THRESHOLD` at 56, raised from 53. This branch
  measures 56.21 percent, and the merge result measures 58.92 percent. The extra
  coverage belongs to other pull requests, so the floor tracks the lower value.
  A later revert on `main` then cannot fail this gate for a reason that belongs
  to another change.
- Corrected the stale comment blocks at both sites. The old text claimed 334
  tests and 53.91 percent, and both numbers were out of date.

The commit changes those values in `ci.yml` and nothing else in that file.

### Which assertions changed, and why the new ones are correct

| Assertion | Before | After |
| - | - | - |
| `Organization.__tablename__` | `"organizations"` | `"orgs"` |
| `SyncLedgerEntry.__tablename__` | `"sync_ledger_entries"` | `"sync_ledger"` |

Both old values named the table that migration `0001` created. Both tests failed
on the base commit, before any change in the pull request. The `ci.yml` comment
recorded that same fact and deselected them for it.

The new values are correct, because the ORM models are the schema owner. Every
route, worker, and API schema reads the ORM names. `health.py` writes
`Organization(org_id=...)` against `orgs`. Migration
`0003_align_schema_with_orm` now builds `orgs` and `sync_ledger`, so the
database, the models, and these two tests all agree.

No assertion was weakened. No test was deleted or skipped.

### The gate result after the fix

```
Collected 398 tests
366 passed, 32 skipped in 5.92s
Required test coverage of 56% reached. Total coverage: 58.92%
```

`pytest (ops platform)` reports success. Zero tests are deselected now.

### Remaining work

The four items from my earlier comment stand, except that item 4 is closed. The
label went on after I cleared `GH_TOKEN` and switched to the correct account,
which is the workaround that issue #1893 describes.

Item 1 of that list also changes. This pull request now removes the deselect
list, which issue #1897 tracked. The continuous integration check that starts
PostgreSQL and runs `alembic check` is still open, and the missing `alembic.ini`
still blocks it.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -75,15 +75,16 @@ env:
   INTERROGATE_PATHS: 'src/ MistHelper.py wsgi.py starlink_dashboard.py web_portal tools'
   COVERAGE_THRESHOLD: ${{ inputs.coverage-threshold || '80' }}
   # Issue #1895: the ops platform suite keeps its own denominator and its own
-  # floor, so it can never move the root percentage. The suite measures 53.91
-  # percent. The floor sits just under that number, so a coverage drop fails the
-  # gate. Issue #1883 unblocks the last 2 tests and raises the floor further.
-  OPS_PLATFORM_COVERAGE_THRESHOLD: '53'
+  # floor, so it can never move the root percentage. Issue #1883 unified the
+  # schema and unblocked the last 2 tests, so no test stays deselected now.
+  # The suite measures 56.18 percent on the runner. The floor sits just under
+  # that number, so a coverage drop fails the gate.
+  OPS_PLATFORM_COVERAGE_THRESHOLD: '56'
   # The count floor proves the gate ran the tests. A broken import path drops a
   # whole module from collection, and this floor turns that drop into a red gate.
-  # The suite collects 334 tests. If you delete a test on purpose, lower this
+  # The suite collects 371 tests. If you delete a test on purpose, lower this
   # number in the same commit.
-  OPS_PLATFORM_MIN_TESTS: '330'
+  OPS_PLATFORM_MIN_TESTS: '365'
   PYLINT_THRESHOLD: ${{ inputs.pylint-threshold || '9.5' }}
   INTERROGATE_THRESHOLD: ${{ inputs.interrogate-threshold || '90' }}
   # Vulture reports 0 findings at 90, at 80, and at 70. The cliff sits at 60, which
@@ -600,18 +601,16 @@ jobs:
         # measures mist-ops-platform/src only, so neither percentage moves the
         # other.
         #
-        # The 2 deselected tests fail on main today. They assert the table names
-        # that migration 0001 creates, and the ORM models declare different names
-        # and different columns. Issue #1883 covers that divergence. The repair is
-        # a schema migration and not a test edit, so the pair stays deselected.
-        # Issue #1897 tracks the removal of this list.
-        # Do not add a test to this list to make a new failure go away. Fix the
-        # code instead.
+        # This job deselected 2 tests until issue #1883. They asserted the table
+        # names that migration 0001 creates, and the ORM models declared other
+        # names. Migration 0003_align_schema_with_orm unified the two schemas on
+        # the ORM models, so both tests now pass and the deselect list is empty.
+        # Issue #1897 tracked that removal.
+        # Do not add a test to a deselect list to make a new failure go away.
+        # Fix the code instead.
         run: |
           pytest --cov=src \
-            --cov-fail-under=${{ env.OPS_PLATFORM_COVERAGE_THRESHOLD }} \
-            --deselect tests/unit/shared/test_models_inventory.py::TestOrganization::test_tablename \
-            --deselect tests/unit/shared/test_models_inventory.py::TestSyncLedgerEntry::test_tablename
+            --cov-fail-under=${{ env.OPS_PLATFORM_COVERAGE_THRESHOLD }}
 
   # ──────────────────────────────────────────────────────────────
   # FAILURE → GITHUB ISSUES (auto-create issues for each failure)

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -77,14 +77,16 @@ env:
   # Issue #1895: the ops platform suite keeps its own denominator and its own
   # floor, so it can never move the root percentage. Issue #1883 unified the
   # schema and unblocked the last 2 tests, so no test stays deselected now.
-  # The suite measures 56.18 percent on the runner. The floor sits just under
-  # that number, so a coverage drop fails the gate.
+  # This branch measures 56.21 percent, and the merge result measures 58.92
+  # percent, because main gained tests from other pull requests. The floor
+  # tracks the lower value, so a later revert on main cannot fail this gate for
+  # a reason that belongs to another change.
   OPS_PLATFORM_COVERAGE_THRESHOLD: '56'
   # The count floor proves the gate ran the tests. A broken import path drops a
   # whole module from collection, and this floor turns that drop into a red gate.
-  # The suite collects 371 tests. If you delete a test on purpose, lower this
+  # The suite collects 398 tests. If you delete a test on purpose, lower this
   # number in the same commit.
-  OPS_PLATFORM_MIN_TESTS: '365'
+  OPS_PLATFORM_MIN_TESTS: '390'
   PYLINT_THRESHOLD: ${{ inputs.pylint-threshold || '9.5' }}
   INTERROGATE_THRESHOLD: ${{ inputs.interrogate-threshold || '90' }}
   # Vulture reports 0 findings at 90, at 80, and at 70. The cliff sits at 60, which

--- a/mist-ops-platform/docs/architecture.md
+++ b/mist-ops-platform/docs/architecture.md
@@ -27,7 +27,7 @@ Juniper Mist Cloud networks.
 
 ### 3. Data Layer
 
-- **PostgreSQL 16**: Primary store with hash-partitioned tables (16 partitions)
+- **PostgreSQL 16**: Primary store with LIST-partitioned tables (one partition per org)
   for `config_revisions`, `device_status_snapshots`, `audit_records`
 - **Redis 7**: Caching, rate limiting, Celery broker
 - **MinIO**: S3-compatible object storage for audit artifacts
@@ -71,7 +71,7 @@ Juniper Mist Cloud networks.
 ## Key Design Decisions
 
 - **R-01**: All Mist API calls isolated to Celery workers (never in API request path)
-- **R-02**: Hash-partitioned tables by org_id for multi-tenant scalability
+- **R-02**: LIST-partitioned tables by org_id for multi-tenant scalability
 - **R-04**: Temporal queries use `captured_at <= timestamp ORDER BY desc LIMIT 1`
 - **R-05**: deepdiff 8.x for JSON config diffing
 - **R-06**: Saga pattern for multi-step rollback

--- a/mist-ops-platform/docs/operations.md
+++ b/mist-ops-platform/docs/operations.md
@@ -19,6 +19,11 @@ docker compose --profile full up -d
 # 3. Run database migrations
 docker compose exec api alembic upgrade head
 
+# The API does not create a table on startup. Alembic owns the schema.
+# Run step 3 before the first request, or every route fails. See issue #1883.
+# Warning: revision 0003_align_schema_with_orm drops every platform table and
+# builds it again. Take a database backup before you run this step.
+
 # 4. Verify health
 curl http://localhost:8000/api/v1/healthz
 

--- a/mist-ops-platform/migrations/versions/0003_align_schema_with_orm.py
+++ b/mist-ops-platform/migrations/versions/0003_align_schema_with_orm.py
@@ -1,0 +1,399 @@
+"""Rebuild the schema so the database matches the ORM models.
+
+Revision ID: 0003_align_schema_with_orm
+Revises: 0002_add_device_uptime_last_seen
+Create Date: 2026-08-24
+
+Issue #1883 reports two schema owners that disagree. The migration `0001_initial`
+built one schema. The ORM models in `src/shared/models/` declare another schema.
+The disagreement broke the login route and every configuration route.
+
+The ORM models win, because every route, worker, and API schema reads the ORM
+column names. The migration `0002` already moved the migration toward the models
+when it added `devices.uptime` and `devices.last_seen_at`.
+
+This migration drops the tables of the old schema and builds the ORM schema. The
+old tables hold no usable data, because no route could read or write them.
+
+Warning: this migration drops every platform table. The drop is irreversible.
+Take a database backup before you run `alembic upgrade head`.
+"""
+
+from __future__ import annotations
+
+import logging
+
+import sqlalchemy as sa
+from alembic import op
+
+revision: str = "0003_align_schema_with_orm"
+down_revision: str = "0002_add_device_uptime_last_seen"
+branch_labels: str | None = None
+depends_on: str | None = None
+
+logger = logging.getLogger(__name__)
+
+
+# Every table of the old migration schema and of the ORM schema.
+DROP_TABLES: tuple[str, ...] = (
+    "audit_records",
+    "baselines",
+    "change_templates",
+    "compliance_audit_packs",
+    "config_revisions",
+    "device_status_snapshots",
+    "devices",
+    "drift_alerts",
+    "golden_images",
+    "incident_change_correlations",
+    "job_checkpoints",
+    "msps",
+    "network_policies",
+    "notification_channels",
+    "organizations",
+    "orgs",
+    "rollout_plans",
+    "rollout_waves",
+    "scheduled_jobs",
+    "sites",
+    "sync_ledger",
+    "sync_ledger_entries",
+    "webhook_envelopes",
+)
+
+# The frozen DDL of every ORM table and index, in dependency order.
+CREATE_STATEMENTS: tuple[str, ...] = (
+    """CREATE TABLE msps (
+    msp_id UUID NOT NULL,
+    name TEXT NOT NULL,
+    api_host TEXT NOT NULL,
+    auth_method VARCHAR(20) NOT NULL,
+    last_sync_at TIMESTAMP WITH TIME ZONE,
+    sync_enabled BOOLEAN DEFAULT 'true' NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (msp_id)
+)""",
+    """CREATE TABLE orgs (
+    org_id UUID NOT NULL,
+    msp_id UUID,
+    name TEXT NOT NULL,
+    api_host TEXT NOT NULL,
+    last_sync_at TIMESTAMP WITH TIME ZONE,
+    sync_enabled BOOLEAN DEFAULT 'true' NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (org_id),
+    FOREIGN KEY(msp_id) REFERENCES msps (msp_id)
+)""",
+    """CREATE TABLE audit_records (
+    record_id BIGSERIAL NOT NULL,
+    org_id UUID NOT NULL,
+    timestamp TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    actor TEXT NOT NULL,
+    entity_type VARCHAR(30) NOT NULL,
+    entity_id UUID NOT NULL,
+    change_type VARCHAR(20) NOT NULL,
+    old_values JSONB,
+    new_values JSONB,
+    revision_id BIGINT,
+    job_id UUID,
+    PRIMARY KEY (record_id, org_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)
+ PARTITION BY LIST (org_id)""",
+    """CREATE TABLE baselines (
+    baseline_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    entity_type VARCHAR(30) NOT NULL,
+    entity_scope UUID NOT NULL,
+    config_payload JSONB NOT NULL,
+    updated_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    updated_by TEXT NOT NULL,
+    PRIMARY KEY (baseline_id),
+    CONSTRAINT uq_baseline_scope UNIQUE (org_id, entity_type, entity_scope),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_baselines_org_id ON baselines (org_id)""",
+    """CREATE TABLE change_templates (
+    template_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    name TEXT NOT NULL,
+    category VARCHAR(30) NOT NULL,
+    parameter_schema JSONB NOT NULL,
+    config_template JSONB NOT NULL,
+    target_entity_type VARCHAR(30) NOT NULL,
+    approval_required BOOLEAN DEFAULT 'false' NOT NULL,
+    author TEXT NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (template_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_change_templates_org_id ON change_templates (org_id)""",
+    """CREATE TABLE compliance_audit_packs (
+    pack_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    framework VARCHAR(20) NOT NULL,
+    date_range_start TIMESTAMP WITH TIME ZONE NOT NULL,
+    date_range_end TIMESTAMP WITH TIME ZONE NOT NULL,
+    included_records JSONB NOT NULL,
+    artifact_url TEXT,
+    export_format VARCHAR(10) DEFAULT 'json' NOT NULL,
+    generated_by TEXT NOT NULL,
+    generated_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (pack_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_compliance_audit_packs_org_id ON compliance_audit_packs (org_id)""",
+    """CREATE TABLE config_revisions (
+    revision_id BIGSERIAL NOT NULL,
+    org_id UUID NOT NULL,
+    entity_type VARCHAR(30) NOT NULL,
+    entity_id UUID NOT NULL,
+    captured_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    content_hash TEXT NOT NULL,
+    config_payload JSONB NOT NULL,
+    actor TEXT,
+    source VARCHAR(20) DEFAULT 'sync' NOT NULL,
+    PRIMARY KEY (revision_id, org_id),
+    CONSTRAINT uq_revision_dedup UNIQUE (entity_id, content_hash, org_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)
+ PARTITION BY LIST (org_id)""",
+    """CREATE TABLE golden_images (
+    image_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    image_type VARCHAR(30) NOT NULL,
+    device_model TEXT NOT NULL,
+    version TEXT NOT NULL,
+    lifecycle_state VARCHAR(20) DEFAULT 'draft' NOT NULL,
+    content_hash TEXT NOT NULL,
+    artifact_url TEXT,
+    approved_by TEXT,
+    approved_at TIMESTAMP WITH TIME ZONE,
+    created_by TEXT NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (image_id),
+    CONSTRAINT uq_golden_image UNIQUE (org_id, image_type, device_model, version),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_golden_images_org_id ON golden_images (org_id)""",
+    """CREATE TABLE incident_change_correlations (
+    correlation_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    incident_type VARCHAR(30) NOT NULL,
+    incident_id TEXT NOT NULL,
+    incident_at TIMESTAMP WITH TIME ZONE NOT NULL,
+    change_revision_id INTEGER,
+    change_job_id UUID,
+    confidence_score FLOAT NOT NULL,
+    detection_method VARCHAR(20) NOT NULL,
+    detected_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (correlation_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_incident_change_correlations_org_id
+ON incident_change_correlations (org_id)""",
+    """CREATE TABLE network_policies (
+    policy_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    mist_entity_id UUID NOT NULL,
+    policy_type VARCHAR(30) NOT NULL,
+    name TEXT NOT NULL,
+    lifecycle_state VARCHAR(20) DEFAULT 'active' NOT NULL,
+    version INTEGER DEFAULT '1' NOT NULL,
+    effective_from TIMESTAMP WITH TIME ZONE,
+    expires_at TIMESTAMP WITH TIME ZONE,
+    dependencies JSONB,
+    last_reviewed_at TIMESTAMP WITH TIME ZONE,
+    reviewed_by TEXT,
+    PRIMARY KEY (policy_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_network_policies_org_id ON network_policies (org_id)""",
+    """CREATE TABLE notification_channels (
+    channel_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    channel_type VARCHAR(20) NOT NULL,
+    name TEXT NOT NULL,
+    destination TEXT NOT NULL,
+    alert_subscriptions TEXT[] NOT NULL,
+    enabled BOOLEAN DEFAULT 'true' NOT NULL,
+    auth_config JSONB,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (channel_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_notification_channels_org_id ON notification_channels (org_id)""",
+    """CREATE TABLE rollout_plans (
+    plan_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    name TEXT NOT NULL,
+    promotion_mode VARCHAR(20) DEFAULT 'manual' NOT NULL,
+    health_gate_criteria JSONB NOT NULL,
+    status VARCHAR(30) DEFAULT 'draft' NOT NULL,
+    created_by TEXT NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (plan_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_rollout_plans_org_id ON rollout_plans (org_id)""",
+    """CREATE TABLE sites (
+    site_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    name TEXT NOT NULL,
+    address TEXT,
+    location JSONB,
+    last_sync_at TIMESTAMP WITH TIME ZONE,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (site_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_sites_org_id ON sites (org_id)""",
+    """CREATE TABLE sync_ledger (
+    id BIGSERIAL NOT NULL,
+    org_id UUID NOT NULL,
+    job_type VARCHAR(30) NOT NULL,
+    started_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    ended_at TIMESTAMP WITH TIME ZONE,
+    status VARCHAR(20) DEFAULT 'running' NOT NULL,
+    rows_affected INTEGER,
+    error_text TEXT,
+    PRIMARY KEY (id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id)
+)""",
+    """CREATE INDEX ix_sync_ledger_org_id ON sync_ledger (org_id)""",
+    """CREATE TABLE webhook_envelopes (
+    id BIGSERIAL NOT NULL,
+    org_id UUID NOT NULL,
+    event_id TEXT NOT NULL,
+    received_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    event_type VARCHAR(50) NOT NULL,
+    payload JSONB NOT NULL,
+    processed_at TIMESTAMP WITH TIME ZONE,
+    status VARCHAR(20) DEFAULT 'pending' NOT NULL,
+    PRIMARY KEY (id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id),
+    UNIQUE (event_id)
+)""",
+    """CREATE INDEX ix_webhook_envelopes_org_id ON webhook_envelopes (org_id)""",
+    """CREATE TABLE devices (
+    device_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    site_id UUID,
+    name TEXT,
+    serial TEXT NOT NULL,
+    model TEXT NOT NULL,
+    device_type VARCHAR(20) NOT NULL,
+    firmware_version TEXT,
+    status VARCHAR(20) DEFAULT 'unknown' NOT NULL,
+    mac_address TEXT,
+    ip_address TEXT,
+    uptime INTEGER,
+    last_seen_at TIMESTAMP WITH TIME ZONE,
+    last_sync_at TIMESTAMP WITH TIME ZONE,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (device_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id),
+    FOREIGN KEY(site_id) REFERENCES sites (site_id),
+    UNIQUE (serial)
+)""",
+    """CREATE INDEX ix_devices_org_id ON devices (org_id)""",
+    """CREATE INDEX ix_devices_site_id ON devices (site_id)""",
+    """CREATE TABLE rollout_waves (
+    plan_id UUID NOT NULL,
+    wave_number INTEGER NOT NULL,
+    target_entities JSONB NOT NULL,
+    status VARCHAR(20) DEFAULT 'pending' NOT NULL,
+    started_at TIMESTAMP WITH TIME ZONE,
+    completed_at TIMESTAMP WITH TIME ZONE,
+    health_check_result JSONB,
+    PRIMARY KEY (plan_id, wave_number),
+    FOREIGN KEY(plan_id) REFERENCES rollout_plans (plan_id)
+)""",
+    """CREATE TABLE scheduled_jobs (
+    job_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    target_entities JSONB NOT NULL,
+    change_payload JSONB NOT NULL,
+    scheduled_at TIMESTAMP WITH TIME ZONE NOT NULL,
+    status VARCHAR(30) DEFAULT 'pending' NOT NULL,
+    pre_check_defs JSONB,
+    post_check_defs JSONB,
+    pre_check_result JSONB,
+    post_check_result JSONB,
+    created_by TEXT NOT NULL,
+    approved_by TEXT,
+    rollout_plan_id UUID,
+    started_at TIMESTAMP WITH TIME ZONE,
+    completed_at TIMESTAMP WITH TIME ZONE,
+    error_message TEXT,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (job_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id),
+    FOREIGN KEY(rollout_plan_id) REFERENCES rollout_plans (plan_id)
+)""",
+    """CREATE INDEX ix_scheduled_jobs_org_id ON scheduled_jobs (org_id)""",
+    """CREATE TABLE device_status_snapshots (
+    snapshot_id BIGSERIAL NOT NULL,
+    org_id UUID NOT NULL,
+    device_id UUID NOT NULL,
+    captured_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    status VARCHAR(20) NOT NULL,
+    port_states JSONB,
+    client_count INTEGER,
+    health_metrics JSONB,
+    PRIMARY KEY (snapshot_id, org_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id),
+    FOREIGN KEY(device_id) REFERENCES devices (device_id)
+)
+ PARTITION BY LIST (org_id)""",
+    """CREATE TABLE drift_alerts (
+    alert_id UUID NOT NULL,
+    org_id UUID NOT NULL,
+    baseline_id UUID NOT NULL,
+    device_id UUID NOT NULL,
+    detected_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    diff_payload JSONB NOT NULL,
+    status VARCHAR(20) DEFAULT 'open' NOT NULL,
+    resolved_at TIMESTAMP WITH TIME ZONE,
+    resolved_by TEXT,
+    PRIMARY KEY (alert_id),
+    FOREIGN KEY(org_id) REFERENCES orgs (org_id),
+    FOREIGN KEY(baseline_id) REFERENCES baselines (baseline_id),
+    FOREIGN KEY(device_id) REFERENCES devices (device_id)
+)""",
+    """CREATE INDEX ix_drift_alerts_org_id ON drift_alerts (org_id)""",
+    """CREATE TABLE job_checkpoints (
+    checkpoint_id BIGSERIAL NOT NULL,
+    job_id UUID NOT NULL,
+    entity_id UUID NOT NULL,
+    step VARCHAR(30) NOT NULL,
+    status VARCHAR(20) NOT NULL,
+    payload JSONB,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL,
+    PRIMARY KEY (checkpoint_id, job_id),
+    FOREIGN KEY(job_id) REFERENCES scheduled_jobs (job_id)
+)""",
+)
+
+
+def upgrade() -> None:
+    """Drop the old schema and build the schema that the ORM models declare."""
+    logger.info("Schema rebuild starts for %d tables.", len(CREATE_STATEMENTS))
+    for table_name in DROP_TABLES:  # Clear both schemas, so no stale column survives.
+        # WHY: CASCADE removes the hash partition children and the foreign keys too.
+        op.execute(sa.text(f"DROP TABLE IF EXISTS {table_name} CASCADE"))
+    for statement in CREATE_STATEMENTS:  # Build in dependency order, so each key resolves.
+        op.execute(sa.text(statement))  # Send one frozen DDL statement to the database.
+    logger.debug("Schema rebuild done for %d tables.", len(CREATE_STATEMENTS))
+
+
+def downgrade() -> None:
+    """Refuse the downgrade, because the old schema held no usable data."""
+    logger.info("Downgrade request starts for revision %s.", revision)
+    # WHY: a downgrade would restore a schema that no route can read. A clear stop
+    # protects the operator better than a silent rebuild of a broken schema.
+    raise NotImplementedError(
+        "Revision 0003_align_schema_with_orm has no downgrade. "
+        "The schema it replaced could not serve a request. "
+        "Restore a database backup instead."
+    )

--- a/mist-ops-platform/requirements-dev.txt
+++ b/mist-ops-platform/requirements-dev.txt
@@ -19,6 +19,14 @@ sqlalchemy>=2.0
 pydantic>=2.10
 pydantic-settings>=2.7
 
+# alembic backs migrations/versions/. tests/unit/shared/test_schema_agreement.py
+# imports alembic at module scope, because it runs the whole migration chain in
+# the Alembic offline mode and compares the rendered DDL against the ORM
+# metadata. Its absence caused a collection error on pull request #2031.
+# pyproject.toml already lists alembic as a runtime dependency, so this pin adds
+# no new package to the project.
+alembic>=1.14
+
 # httpx serves both the application client and the FastAPI test client.
 # The CI run for issue #1895 proved which name each layer needs.
 #   - src/ imports httpx directly for the application client.

--- a/mist-ops-platform/src/api/main.py
+++ b/mist-ops-platform/src/api/main.py
@@ -18,9 +18,10 @@ async def lifespan(app: FastAPI) -> AsyncIterator[None]:
     engine = build_engine(settings)
     app.state.engine = engine
 
-    from src.shared.db import ensure_tables_exist
-
-    await ensure_tables_exist(engine)
+    # WHY: Alembic owns the schema. See issue #1883. A `create_all` call on
+    # startup used `checkfirst`, so it skipped every table the migration built
+    # and it added the missing ORM tables. The result matched neither owner.
+    # Run `alembic upgrade head` before you start the API. See docs/operations.md.
 
     yield
     await engine.dispose()

--- a/mist-ops-platform/src/shared/db.py
+++ b/mist-ops-platform/src/shared/db.py
@@ -62,10 +62,18 @@ async def get_session(
 
 
 async def ensure_tables_exist(engine) -> None:  # noqa: ANN001
-    """Create all ORM tables on startup. Idempotent via checkfirst.
+    """Create all ORM tables. Use this helper in a test fixture only.
+
+    Warning: do not call this helper on the API startup path. `create_all` uses
+    `checkfirst`, so it skips a table that already exists and it never alters
+    one. It therefore cannot repair a schema difference, and it hides the
+    difference instead of reporting it. See issue #1883.
+
+    Alembic owns the deployment schema. Run `alembic upgrade head` before you
+    start the API. See docs/operations.md.
 
     Uses an advisory lock to prevent race conditions when multiple
-    uvicorn workers start simultaneously.
+    test workers start simultaneously.
     """
     import src.shared.models  # noqa: F401 — register all models
     from src.shared.models.base import Base
@@ -80,7 +88,12 @@ async def ensure_tables_exist(engine) -> None:  # noqa: ANN001
 
 
 async def ensure_org_partitions(engine, org_id: str) -> None:  # noqa: ANN001
-    """Create LIST partitions for one org across all partitioned tables."""
+    """Create LIST partitions for one org across all partitioned tables.
+
+    The three tables in `PARTITIONED_TABLES` use `PARTITION BY LIST (org_id)`.
+    The models declare that strategy, and migration `0003_align_schema_with_orm`
+    builds it. A LIST bound therefore matches the parent table.
+    """
     validated = uuid_mod.UUID(str(org_id))
     safe_suffix = validated.hex
 

--- a/mist-ops-platform/src/shared/models/config.py
+++ b/mist-ops-platform/src/shared/models/config.py
@@ -27,7 +27,7 @@ from src.shared.models.base import Base
 
 
 # ---------------------------------------------------------------------------
-# E-04: ConfigRevision (hash-partitioned by org_id)
+# E-04: ConfigRevision (LIST-partitioned by org_id)
 # ---------------------------------------------------------------------------
 class ConfigRevision(Base):
     """Immutable configuration snapshot captured from Mist API."""
@@ -70,7 +70,7 @@ class ConfigRevision(Base):
 
 
 # ---------------------------------------------------------------------------
-# E-05: DeviceStatusSnapshot (hash-partitioned by org_id)
+# E-05: DeviceStatusSnapshot (LIST-partitioned by org_id)
 # ---------------------------------------------------------------------------
 class DeviceStatusSnapshot(Base):
     """Point-in-time device status for time-travel queries."""

--- a/mist-ops-platform/src/shared/models/operations.py
+++ b/mist-ops-platform/src/shared/models/operations.py
@@ -132,7 +132,7 @@ class JobCheckpoint(Base):
 
 
 # ---------------------------------------------------------------------------
-# E-06: AuditRecord (hash-partitioned by org_id)
+# E-06: AuditRecord (LIST-partitioned by org_id)
 # ---------------------------------------------------------------------------
 class AuditRecord(Base):
     """Field-level change audit trail entry."""

--- a/mist-ops-platform/tests/unit/shared/test_models_inventory.py
+++ b/mist-ops-platform/tests/unit/shared/test_models_inventory.py
@@ -44,7 +44,10 @@ class TestOrganization:
     """Verify Organization entity."""
 
     def test_tablename(self) -> None:
-        assert Organization.__tablename__ == "organizations"
+        # Issue #1883: this test asserted "organizations", which is the name that
+        # migration 0001 used. Every route and worker reads the ORM name "orgs",
+        # so the ORM name wins. Migration 0003 now builds "orgs".
+        assert Organization.__tablename__ == "orgs"
 
     def test_primary_key(self) -> None:
         cols = {c.name for c in Organization.__table__.primary_key.columns}
@@ -94,7 +97,10 @@ class TestSyncLedgerEntry:
     """Verify SyncLedgerEntry entity."""
 
     def test_tablename(self) -> None:
-        assert SyncLedgerEntry.__tablename__ == "sync_ledger_entries"
+        # Issue #1883: this test asserted "sync_ledger_entries", which is the name
+        # that migration 0001 used. The ORM names the table "sync_ledger", and the
+        # ORM is the schema owner. Migration 0003 now builds "sync_ledger".
+        assert SyncLedgerEntry.__tablename__ == "sync_ledger"
 
     def test_primary_key(self) -> None:
         cols = {c.name for c in SyncLedgerEntry.__table__.primary_key.columns}

--- a/mist-ops-platform/tests/unit/shared/test_schema_agreement.py
+++ b/mist-ops-platform/tests/unit/shared/test_schema_agreement.py
@@ -1,0 +1,206 @@
+"""Prove that the Alembic migration chain and the ORM models build one schema.
+
+Issue #1883 records four disagreements between
+`migrations/versions/0001_initial.py` and the models in `src/shared/models/`.
+The disagreements break the login route and every configuration route.
+
+This module runs the whole migration chain in the Alembic offline mode. The
+offline mode renders the DDL as text and needs no PostgreSQL server. The tests
+then read that DDL and compare it against the ORM metadata.
+"""
+
+from __future__ import annotations
+
+import importlib.util
+import io
+import logging
+import re
+import sys
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import pytest
+from alembic.migration import MigrationContext
+from alembic.operations import Operations
+
+from src.shared.db import PARTITIONED_TABLES
+from src.shared.models.base import Base
+
+if TYPE_CHECKING:  # The module type is an annotation only, so keep it out of runtime.
+    from types import ModuleType
+
+logger = logging.getLogger(__name__)
+
+# The migration files live beside the package root, so walk up to the project.
+VERSIONS_DIR = Path(__file__).resolve().parents[3] / "migrations" / "versions"
+
+# A hash partition child carries a `_pNN` suffix. A LIST partition child carries
+# a `_org_<hex>` suffix. Neither child is an ORM table, so the tests skip both.
+PARTITION_CHILD_PATTERN = re.compile(r"_p\d+$|_org_[0-9a-f]{32}$")
+
+# A statement ends at a semicolon that closes the line in the Alembic output.
+STATEMENT_SPLIT_PATTERN = re.compile(r";\s*\n")
+
+# `CREATE TABLE name (` and `DROP TABLE name` both name the table in group 1.
+CREATE_TABLE_PATTERN = re.compile(
+    r"CREATE\s+TABLE\s+(?:IF\s+NOT\s+EXISTS\s+)?([A-Za-z_][A-Za-z0-9_]*)",
+    re.IGNORECASE,
+)
+DROP_TABLE_PATTERN = re.compile(
+    r"DROP\s+TABLE\s+(?:IF\s+EXISTS\s+)?([A-Za-z_][A-Za-z0-9_]*)",
+    re.IGNORECASE,
+)
+
+
+def _load_migration_modules() -> list[ModuleType]:
+    """Import every migration file and return the modules in upgrade order."""
+    logger.info("Migration module load starts from %s.", VERSIONS_DIR)
+    modules: dict[str, ModuleType] = {}  # Map a revision identifier to its module.
+    for path in sorted(VERSIONS_DIR.glob("*.py")):  # Sort so the load order is stable.
+        name = f"_mist_ops_migration_{path.stem}"  # Use a private name to avoid a clash.
+        spec = importlib.util.spec_from_file_location(name, path)  # Build the import spec.
+        if spec is None or spec.loader is None:  # A missing loader means the file is unreadable.
+            pytest.fail(f"The migration file {path} is not importable.")
+        module = importlib.util.module_from_spec(spec)  # Create the empty module object.
+        sys.modules[name] = module  # Register the module, because `from __future__` needs it.
+        spec.loader.exec_module(module)  # Run the file, which defines `upgrade` and the ids.
+        modules[module.revision] = module  # Key the module by the revision it declares.
+    ordered = _order_by_revision_chain(modules)  # Follow `down_revision` to get the real order.
+    logger.debug("Migration module load done for %d modules.", len(ordered))
+    return ordered
+
+
+def _order_by_revision_chain(modules: dict[str, ModuleType]) -> list[ModuleType]:
+    """Walk the `down_revision` chain and return the modules in upgrade order."""
+    logger.info("Revision chain walk starts for %d modules.", len(modules))
+    children = {m.down_revision: m for m in modules.values()}  # Map a parent to its child.
+    ordered: list[ModuleType] = []  # Collect the modules from the root to the head.
+    cursor: str | None = None  # The root migration declares `down_revision = None`.
+    while cursor in children:  # Stop when no migration names the cursor as its parent.
+        module = children[cursor]  # Take the child of the current revision.
+        ordered.append(module)  # Append it, because it is the next upgrade step.
+        cursor = module.revision  # Move the cursor to the revision that just ran.
+    if len(ordered) != len(modules):  # A gap means a broken or a branched chain.
+        pytest.fail("The migration chain is broken. Check every `down_revision` value.")
+    logger.debug("Revision chain walk done for %d modules.", len(ordered))
+    return ordered
+
+
+def _render_migration_ddl() -> str:
+    """Run every migration in the offline mode and return the rendered DDL."""
+    logger.info("Offline migration render starts.")
+    buffer = io.StringIO()  # The offline mode writes the DDL into this buffer.
+    context = MigrationContext.configure(  # Build a context that renders instead of executes.
+        dialect_name="postgresql",  # Render for PostgreSQL, because that is the target.
+        opts={"as_sql": True, "output_buffer": buffer},  # `as_sql` selects the offline mode.
+    )
+    with Operations.context(context):  # Install the `alembic.op` proxy for the migrations.
+        for module in _load_migration_modules():  # Run the chain from the root to the head.
+            module.upgrade()  # Emit the DDL of this one migration into the buffer.
+    rendered = buffer.getvalue()  # Read the whole rendered script.
+    logger.debug("Offline migration render done with %d characters.", len(rendered))
+    return rendered
+
+
+def _statements(ddl: str) -> list[str]:
+    """Split the rendered DDL into single statements."""
+    logger.info("Statement split starts for %d characters.", len(ddl))
+    parts = [part.strip() for part in STATEMENT_SPLIT_PATTERN.split(ddl)]  # Cut at each end.
+    kept = [part for part in parts if part]  # Drop the empty tail that the final cut leaves.
+    logger.debug("Statement split done with %d statements.", len(kept))
+    return kept
+
+
+def _surviving_tables(ddl: str) -> dict[str, str]:
+    """Return the tables the migration chain leaves behind, keyed to their DDL."""
+    logger.info("Surviving table scan starts.")
+    tables: dict[str, str] = {}  # Hold the last `CREATE TABLE` text for each table name.
+    for statement in _statements(ddl):  # Replay the script in order, so the last write wins.
+        dropped = DROP_TABLE_PATTERN.match(statement)  # A drop removes the table again.
+        if dropped is not None:  # Handle the drop first, because a drop never creates.
+            tables.pop(dropped.group(1), None)  # Forget the table, and ignore an absent name.
+            continue  # Move on, because a drop holds no column text.
+        created = CREATE_TABLE_PATTERN.match(statement)  # A create adds or replaces the table.
+        if created is not None:  # Record only a statement that starts with `CREATE TABLE`.
+            tables[created.group(1)] = statement  # Store the text for the column assertions.
+    logger.debug("Surviving table scan done with %d tables.", len(tables))
+    return tables
+
+
+def _parent_tables(ddl: str) -> dict[str, str]:
+    """Return the surviving tables without the partition children."""
+    logger.info("Partition child filter starts.")
+    survivors = _surviving_tables(ddl)  # Start from every table the chain leaves behind.
+    parents: dict[str, str] = {}  # Hold only the parent tables of the final schema.
+    for name, statement in survivors.items():  # Read every table the chain leaves behind.
+        if PARTITION_CHILD_PATTERN.search(name) is not None:  # Match a child name suffix.
+            continue  # Skip the child, because no ORM model declares a partition child.
+        parents[name] = statement  # Keep the parent, because the ORM must declare it.
+    logger.debug("Partition child filter done with %d parents.", len(parents))
+    return parents
+
+
+@pytest.fixture(scope="module")
+def migration_ddl() -> str:
+    """Render the migration chain one time for every test in this module."""
+    return _render_migration_ddl()
+
+
+def test_migration_creates_every_orm_table(migration_ddl: str) -> None:
+    """The migration must build every table that the ORM models declare."""
+    parents = _parent_tables(migration_ddl)  # Read the tables the chain leaves behind.
+    expected = set(Base.metadata.tables)  # The ORM metadata is the schema owner.
+    missing = sorted(expected - set(parents))  # A missing table breaks every query on it.
+    assert not missing, f"The migration builds no table for the ORM tables {missing}."
+
+
+def test_migration_leaves_no_table_outside_the_orm(migration_ddl: str) -> None:
+    """The migration must not leave a table that no ORM model declares."""
+    parents = _parent_tables(migration_ddl)  # Read the tables the chain leaves behind.
+    expected = set(Base.metadata.tables)  # The ORM metadata is the schema owner.
+    extra = sorted(set(parents) - expected)  # An extra table is dead weight and a trap.
+    assert not extra, f"The migration leaves the non-ORM tables {extra}."
+
+
+def test_config_revisions_primary_key_matches_the_model(migration_ddl: str) -> None:
+    """The `config_revisions` key must be `revision_id` plus `org_id`."""
+    parents = _parent_tables(migration_ddl)  # Read the tables the chain leaves behind.
+    assert "config_revisions" in parents, "The migration builds no `config_revisions` table."
+    statement = parents["config_revisions"]  # Take the final definition of that one table.
+    assert "revision_id" in statement, "The `config_revisions` table holds no `revision_id`."
+    assert (
+        "PRIMARY KEY (revision_id, org_id)" in statement
+    ), "The `config_revisions` primary key does not match the `ConfigRevision` model."
+
+
+def test_config_revisions_columns_match_the_model(migration_ddl: str) -> None:
+    """Every `ConfigRevision` column must exist in the migrated table."""
+    parents = _parent_tables(migration_ddl)  # Read the tables the chain leaves behind.
+    statement = parents.get("config_revisions", "")  # Take the final definition, or an empty text.
+    model_columns = [c.name for c in Base.metadata.tables["config_revisions"].columns]  # Read ORM.
+    missing = [name for name in model_columns if name not in statement]  # Compare name by name.
+    assert not missing, f"The migrated `config_revisions` table holds no columns {missing}."
+
+
+@pytest.mark.parametrize("table_name", PARTITIONED_TABLES)
+def test_partitioned_tables_use_list_partitioning(migration_ddl: str, table_name: str) -> None:
+    """Each partitioned table must use LIST, because `ensure_org_partitions` sends a LIST bound."""
+    parents = _parent_tables(migration_ddl)  # Read the tables the chain leaves behind.
+    statement = parents.get(table_name, "")  # Take the final definition of this one table.
+    assert statement, f"The migration builds no `{table_name}` table."
+    assert (
+        "PARTITION BY LIST (org_id)" in statement
+    ), f"The `{table_name}` table does not use LIST partitioning by `org_id`."
+    assert "PARTITION BY HASH" not in statement, (
+        f"The `{table_name}` table still uses HASH partitioning. "
+        "A LIST bound fails against a hash-partitioned parent."
+    )
+
+
+def test_model_declares_list_partitioning_for_every_partitioned_table() -> None:
+    """The ORM models must declare the same partition strategy as `PARTITIONED_TABLES`."""
+    for table_name in PARTITIONED_TABLES:  # Check each table the login path partitions.
+        table = Base.metadata.tables[table_name]  # Read the ORM table object by its name.
+        strategy = table.dialect_options["postgresql"].get("partition_by")  # Read the option.
+        message = f"The `{table_name}` model declares the partition strategy {strategy!r}."
+        assert strategy == "LIST (org_id)", message  # A HASH parent rejects a LIST bound.

